# Part 3 of Toronto Neighbourhoods' segmentation and clustering

## This is a sequel to the two previous parts of this project that involves using some techniques to scrape a wikipedia page: https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M.  After the scraping of the required table, it was cleaned up and preprocessed to facilitate the last stage.

## In this part, we give a synopsis of how we obtained the final data that we would work on. Then, we present the dataset for segmenting and clustering of the Toronto Neighbourhoods. We used folium to display the maps of the study.

## We equally, utilized the foursquare to investigate venues within the Neighbourhood of interest, which is Downtown Toronto.

In [173]:
# Essential libraries are imported and installed in this code cell

import numpy as np # library to handle data in a vectorized manner
!pip install folium
!pip install pgeocode
!pip install geopy

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import numpy as np # library to handle data in a vectorized manner

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


### There are several methods one can apply to scrape web pages. Instead of using BeautifulSoup we choose the dataframe approach and the request. The code below was used to extract table which was converted to dataframe.

In [174]:
import requests
import pandas as pd

url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
html = requests.get(url).content
df_list = pd.read_html(html)
df = df_list[0]
print(df)

    Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
7           M8A      Not assigned   
8           M9A         Etobicoke   
9           M1B       Scarborough   
10          M2B      Not assigned   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
15          M7B      Not assigned   
16          M8B      Not assigned   
17          M9B         Etobicoke   
18          M1C       Scarborough   
19          M2C      Not assigned   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
24          M7C      Not assigned   
25          M8C      Not assigned   
2

### The obtained dataframe was to local machine as 'my trial data.csv'.  

In [175]:
df.to_csv('my trial data.csv')

In [176]:
df_1 = pd.read_csv(r'C:\Users\owner\my trial data.csv')
df_1.drop([0, 1, 7, 10, 15, 16, 19, 24, 25, 28, 29, 33, 34, 35, 37, 38, 42, 43, 44, 51, 52, 53, 60, 61, 62, 69, 70, 71, 78, 79, 87, 88, 96, 97, 101, 105, 106, 110, 115, 118, 119, 123, 124, 125, 127, 128, 131, 132, 133, 134, 136, 137, 140, 141, 145, 146, 149, 150, 154, 155, 158, 159,161, 162, 163, 164, 166, 167, 170, 171, 172, 173, 174, 175, 176, 177, 179], inplace = True)
df_1.drop(axis = 0, index = None, columns = 'Unnamed: 0', inplace = True)
df_1 = df_1.reset_index()
df_1.drop(axis =0, columns = 'index', inplace = True)
df_1.head(104)    

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Vic

In [177]:
# Useful libraries that convert addresses to geographic coordinates
import geocoder # import geocoder
import pgeocode
import pandas as pd
import numpy as np

In [178]:
dlist = df_1['Postal Code'].to_list()
print(dlist)

['M3A', 'M4A', 'M5A', 'M6A', 'M7A', 'M9A', 'M1B', 'M3B', 'M4B', 'M5B', 'M6B', 'M9B', 'M1C', 'M3C', 'M4C', 'M5C', 'M6C', 'M9C', 'M1E', 'M4E', 'M5E', 'M6E', 'M1G', 'M4G', 'M5G', 'M6G', 'M1H', 'M2H', 'M3H', 'M4H', 'M5H', 'M6H', 'M1J', 'M2J', 'M3J', 'M4J', 'M5J', 'M6J', 'M1K', 'M2K', 'M3K', 'M4K', 'M5K', 'M6K', 'M1L', 'M2L', 'M3L', 'M4L', 'M5L', 'M6L', 'M9L', 'M1M', 'M2M', 'M3M', 'M4M', 'M5M', 'M6M', 'M9M', 'M1N', 'M2N', 'M3N', 'M4N', 'M5N', 'M6N', 'M9N', 'M1P', 'M2P', 'M4P', 'M5P', 'M6P', 'M9P', 'M1R', 'M2R', 'M4R', 'M5R', 'M6R', 'M7R', 'M9R', 'M1S', 'M4S', 'M5S', 'M6S', 'M1T', 'M4T', 'M5T', 'M1V', 'M4V', 'M5V', 'M8V', 'M9V', 'M1W', 'M4W', 'M5W', 'M8W', 'M9W', 'M1X', 'M4X', 'M5X', 'M8X', 'M4Y', 'M7Y', 'M8Y', 'M8Z']


In [179]:
ps = pgeocode.Nominatim('ca')
ps_2 = ps.query_postal_code(dlist)
ps_2

postal_code country code  \
0           M3A           CA   
1           M4A           CA   
2           M5A           CA   
3           M6A           CA   
4           M7A           CA   
5           M9A           CA   
6           M1B           CA   
7           M3B           CA   
8           M4B           CA   
9           M5B           CA   
10          M6B           CA   
11          M9B           CA   
12          M1C           CA   
13          M3C           CA   
14          M4C           CA   
15          M5C           CA   
16          M6C           CA   
17          M9C           CA   
18          M1E           CA   
19          M4E           CA   
20          M5E           CA   
21          M6E           CA   
22          M1G           CA   
23          M4G           CA   
24          M5G           CA   
25          M6G           CA   
26          M1H           CA   
27          M2H           CA   
28          M3H           CA   
29          M4H           CA   
30          M5H           CA   
31          M6H           CA   
32          M1J           CA   
33          M2J           CA   
34          M3J           CA   
35          M4J           CA   
36          M5J           CA   
37          M6J           CA   
38          M1K           CA   
39          M2K           CA   
40          M3K           CA   
41          M4K           CA   
42          M5K           CA   
43          M6K           CA   
44          M1L           CA   
45          M2L           CA   
46          M3L           CA   
47          M4L           CA   
48          M5L           CA   
49          M6L           CA   
50          M9L           CA   
51          M1M           CA   
52          M2M           CA   
53          M3M           CA   
54          M4M           CA   
55          M5M           CA   
56          M6M           CA   
57          M9M           CA   
58          M1N           CA   
59          M2N           CA   
60          M3N           CA   
61          M4N           CA   
62          M5N           CA   
63          M6N           CA   
64          M9N           CA   
65          M1P           CA   
66          M2P           CA   
67          M4P           CA   
68          M5P           CA   
69          M6P           CA   
70          M9P           CA   
71          M1R           CA   
72          M2R           CA   
73          M4R           CA   
74          M5R           CA   
75          M6R           CA   
76          M7R          NaN   
77          M9R           CA   
78          M1S           CA   
79          M4S           CA   
80          M5S           CA   
81          M6S           CA   
82          M1T           CA   
83          M4T           CA   
84          M5T           CA   
85          M1V           CA   
86          M4V           CA   
87          M5V           CA   
88          M8V           CA   
89          M9V           CA   
90          M1W           CA   
91          M4W           CA   
92          M5W           CA   
93          M8W           CA   
94          M9W           CA   
95          M1X           CA   
96          M4X           CA   
97          M5X           CA   
98          M8X           CA   
99          M4Y           CA   
100         M7Y           CA   
101         M8Y           CA   
102         M8Z           CA   

                                            place_name state_name state_code  \
0    North York (York Heights / Victoria Village / ...    Ontario         ON   
1             North York (Sweeney Park / Wigmore Park)    Ontario         ON   
2     Downtown Toronto (Regent Park / Port of Toronto)    Ontario         ON   
3       North York (Lawrence Manor / Lawrence Heights)    Ontario         ON   
4           Queen's Park Ontario Provincial Government    Ontario         ON   
5                         Etobicoke (Islington Avenue)    Ontario         ON   
6                  Scarborough (Malvern / Rouge River)    Ontario         ON   
7                                  

### We drop some of the colums in the above dataframe and equally fill the NaN values in the latitude and longitude.

In [180]:
ps_2.drop(['country code', 'place_name', 'state_name', 'state_code', 'county_name', 'county_code', 'community_name', 'community_code', 'accuracy'], axis = 1, inplace = True)
ps_2['longitude'] = ps_2['longitude'].fillna(-79.6441)
ps_2['latitude'] = ps_2['latitude'].fillna(43.5890)
ps_2

postal_code  latitude  longitude
0           M3A   43.7545   -79.3300
1           M4A   43.7276   -79.3148
2           M5A   43.6555   -79.3626
3           M6A   43.7223   -79.4504
4           M7A   43.6641   -79.3889
5           M9A   43.6662   -79.5282
6           M1B   43.8113   -79.1930
7           M3B   43.7450   -79.3590
8           M4B   43.7063   -79.3094
9           M5B   43.6572   -79.3783
10          M6B   43.7081   -79.4479
11          M9B   43.6505   -79.5517
12          M1C   43.7878   -79.1564
13          M3C   43.7334   -79.3329
14          M4C   43.6913   -79.3116
15          M5C   43.6513   -79.3756
16          M6C   43.6915   -79.4307
17          M9C   43.6437   -79.5767
18          M1E   43.7678   -79.1866
19          M4E   43.6784   -79.2941
20          M5E   43.6456   -79.3754
21          M6E   43.6889   -79.4507
22          M1G   43.7712   -79.2144
23          M4G   43.7124   -79.3644
24          M5G   43.6564   -79.3860
25          M6G   43.6683   -79.4205
26          M1H   43.7686   -79.2389
27          M2H   43.8015   -79.3577
28          M3H   43.7535   -79.4472
29          M4H   43.7059   -79.3464
30          M5H   43.6496   -79.3833
31          M6H   43.6655   -79.4378
32          M1J   43.7464   -79.2323
33          M2J   43.7801   -79.3479
34          M3J   43.7694   -79.4921
35          M4J   43.6872   -79.3368
36          M5J   43.6230   -79.3936
37          M6J   43.6480   -79.4177
38          M1K   43.7298   -79.2639
39          M2K   43.7797   -79.3813
40          M3K   43.7390   -79.4692
41          M4K   43.6803   -79.3538
42          M5K   43.6469   -79.3823
43          M6K   43.6383   -79.4301
44          M1L   43.7122   -79.2843
45          M2L   43.7547   -79.3764
46          M3L   43.7334   -79.5116
47          M4L   43.6693   -79.3155
48          M5L   43.6492   -79.3823
49          M6L   43.7137   -79.4869
50          M9L   43.7598   -79.5565
51          M1M   43.7247   -79.2312
52          M2M   43.7915   -79.4103
53          M3M   43.7319   -79.4928
54          M4M   43.6561   -79.3406
55          M5M   43.7335   -79.4177
56          M6M   43.6934   -79.4857
57          M9M   43.7366   -79.5401
58          M1N   43.6952   -79.2646
59          M2N   43.7673   -79.4111
60          M3N   43.7568   -79.5210
61          M4N   43.7301   -79.3935
62          M5N   43.7113   -79.4195
63          M6N   43.6748   -79.4839
64          M9N   43.7068   -79.5170
65          M1P   43.7612   -79.2707
66          M2P   43.7500   -79.3978
67          M4P   43.7135   -79.3887
68          M5P   43.6966   -79.4120
69          M6P   43.6605   -79.4633
70          M9P   43.6949   -79.5323
71          M1R   43.7507   -79.3003
72          M2R   43.7786   -79.4450
73          M4R   43.7143   -79.4065
74          M5R   43.6736   -79.4035
75          M6R   43.6469   -79.4521
76          M7R   43.5890   -79.6441
77          M9R   43.6898   -79.5582
78          M1S   43.7946   -79.2644
79          M4S   43.7020   -79.3853
80          M5S   43.6629   -79.3987
81          M6S   43.6512   -79.4828
82          M1T   43.7812   -79.3036
83          M4T   43.6899   -79.3853
84          M5T   43.6541   -79.3978
85          M1V   43.8177   -79.2819
86          M4V   43.6861   -79.4025
87          M5V   43.6404   -79.3995
88          M8V   43.6075   -79.5013
89          M9V   43.7432   -79.5876
90          M1W   43.8016   -79.3216
91          M4W   43.6827   -79.3730
92          M5W   43.6437   -79.3787
93          M8W   43.6021   -79.5402
94          M9W   43.7144   -79.5909
95          M1X   43.8340   -79.2069
96          M4X   43.6684   -79.3689
97          M5X   43.6492   -79.3823
98          M8X   43.6518   -79.5076
99          M4Y   43.6656   -79.3830
100         M7Y   43.7804   -79.2505
101         M8Y   43.6325   -79.4939
102         M8Z   43.6256   -79.5231

In [181]:
ps_2.rename(columns = {"postal_code": "Postal Code", "latitude": "Latitude", "longitude": "Longitude"}, inplace = True)
ps_2

Postal Code  Latitude  Longitude
0           M3A   43.7545   -79.3300
1           M4A   43.7276   -79.3148
2           M5A   43.6555   -79.3626
3           M6A   43.7223   -79.4504
4           M7A   43.6641   -79.3889
5           M9A   43.6662   -79.5282
6           M1B   43.8113   -79.1930
7           M3B   43.7450   -79.3590
8           M4B   43.7063   -79.3094
9           M5B   43.6572   -79.3783
10          M6B   43.7081   -79.4479
11          M9B   43.6505   -79.5517
12          M1C   43.7878   -79.1564
13          M3C   43.7334   -79.3329
14          M4C   43.6913   -79.3116
15          M5C   43.6513   -79.3756
16          M6C   43.6915   -79.4307
17          M9C   43.6437   -79.5767
18          M1E   43.7678   -79.1866
19          M4E   43.6784   -79.2941
20          M5E   43.6456   -79.3754
21          M6E   43.6889   -79.4507
22          M1G   43.7712   -79.2144
23          M4G   43.7124   -79.3644
24          M5G   43.6564   -79.3860
25          M6G   43.6683   -79.4205
26          M1H   43.7686   -79.2389
27          M2H   43.8015   -79.3577
28          M3H   43.7535   -79.4472
29          M4H   43.7059   -79.3464
30          M5H   43.6496   -79.3833
31          M6H   43.6655   -79.4378
32          M1J   43.7464   -79.2323
33          M2J   43.7801   -79.3479
34          M3J   43.7694   -79.4921
35          M4J   43.6872   -79.3368
36          M5J   43.6230   -79.3936
37          M6J   43.6480   -79.4177
38          M1K   43.7298   -79.2639
39          M2K   43.7797   -79.3813
40          M3K   43.7390   -79.4692
41          M4K   43.6803   -79.3538
42          M5K   43.6469   -79.3823
43          M6K   43.6383   -79.4301
44          M1L   43.7122   -79.2843
45          M2L   43.7547   -79.3764
46          M3L   43.7334   -79.5116
47          M4L   43.6693   -79.3155
48          M5L   43.6492   -79.3823
49          M6L   43.7137   -79.4869
50          M9L   43.7598   -79.5565
51          M1M   43.7247   -79.2312
52          M2M   43.7915   -79.4103
53          M3M   43.7319   -79.4928
54          M4M   43.6561   -79.3406
55          M5M   43.7335   -79.4177
56          M6M   43.6934   -79.4857
57          M9M   43.7366   -79.5401
58          M1N   43.6952   -79.2646
59          M2N   43.7673   -79.4111
60          M3N   43.7568   -79.5210
61          M4N   43.7301   -79.3935
62          M5N   43.7113   -79.4195
63          M6N   43.6748   -79.4839
64          M9N   43.7068   -79.5170
65          M1P   43.7612   -79.2707
66          M2P   43.7500   -79.3978
67          M4P   43.7135   -79.3887
68          M5P   43.6966   -79.4120
69          M6P   43.6605   -79.4633
70          M9P   43.6949   -79.5323
71          M1R   43.7507   -79.3003
72          M2R   43.7786   -79.4450
73          M4R   43.7143   -79.4065
74          M5R   43.6736   -79.4035
75          M6R   43.6469   -79.4521
76          M7R   43.5890   -79.6441
77          M9R   43.6898   -79.5582
78          M1S   43.7946   -79.2644
79          M4S   43.7020   -79.3853
80          M5S   43.6629   -79.3987
81          M6S   43.6512   -79.4828
82          M1T   43.7812   -79.3036
83          M4T   43.6899   -79.3853
84          M5T   43.6541   -79.3978
85          M1V   43.8177   -79.2819
86          M4V   43.6861   -79.4025
87          M5V   43.6404   -79.3995
88          M8V   43.6075   -79.5013
89          M9V   43.7432   -79.5876
90          M1W   43.8016   -79.3216
91          M4W   43.6827   -79.3730
92          M5W   43.6437   -79.3787
93          M8W   43.6021   -79.5402
94          M9W   43.7144   -79.5909
95          M1X   43.8340   -79.2069
96          M4X   43.6684   -79.3689
97          M5X   43.6492   -79.3823
98          M8X   43.6518   -79.5076
99          M4Y   43.6656   -79.3830
100         M7Y   43.7804   -79.2505
101         M8Y   43.6325   -79.4939
102         M8Z   43.6256   -79.5231

### The geographic coordinate dataframe ps_2 and Postal code dataframe, df_1 were merged to produce a desired result which was presented in the first part of this project.

In [182]:
df_new = pd.merge(df_1, ps_2, how = 'inner', on = 'Postal Code')
df_new.head(104)

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  Latitude  Longitude  
0                                            Parkwoods   43.7545   -79.3300  

In [183]:
print('The dataset has {} number of Boroughs and {} number of Neighbourhods'.format(len(df_new['Borough'].unique()),len(df_new['Neighbourhood'])))

The dataset has 10 number of Boroughs and 103 number of Neighbourhods


## Lets get the latitude and longitude of city of Toronto to enable mapping of its Neighbourhoods.

In [187]:
address = 'City of Toronto, Ontario'

#geopy.geopcoders.options.default_user_agent = "my-application"

geolocator = Nominatim(user_agent = "DoDo")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


In [188]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighbourhood in zip(df_new['Latitude'], df_new['Longitude'], df_new['Borough'], df_new['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=7,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.5,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

## Analysizing the Neighbourhoods of Toronto, in the state of Ontario.

### From the table above we observe that there are four sectionss of Toronto: Central Toronto, East Toronto, West Toronto, and Downtown Toronto. Downtown seems to have attributes of interest and hence will be the target of cluster of neighbourhoods.

#### The code below will give us the Boroughs that contain Toronto, which shall be object of our study.

In [192]:
boro_toronto = df_new[df_new['Borough'].str.contains('Toronto')].reset_index(drop = True)
boro_toronto

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M4E      East Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M6H      West Toronto   
10         M5J  Downtown Toronto   
11         M6J      West Toronto   
12         M4K      East Toronto   
13         M5K  Downtown Toronto   
14         M6K      West Toronto   
15         M4L      East Toronto   
16         M5L  Downtown Toronto   
17         M4M      East Toronto   
18         M4N   Central Toronto   
19         M5N   Central Toronto   
20         M4P   Central Toronto   
21         M5P   Central Toronto   
22         M6P      West Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
25         M6R      West Toronto   
26         M4S   Central Toronto   
27         M5S  Downtown Toronto   
28         M6S      West Toronto   
29         M4T   Central Toronto   
30         M5T  Downtown Toronto   
31         M4V   Central Toronto   
32         M5V  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   

                                        Neighbourhood  Latitude  Longitude  
0                           Regent Park, Harbourfront   43.6555   -79.3626  
1         Queen's Park, Ontario Provincial Government   43.6641   -79.3889  
2                            Garden District, Ryerson   43.6572   -79.3783  
3                                      St. James Town   43.6513   -79.3756  
4                                         The Beaches   43.6784   -79.2941  
5                                         Berczy Park   43.6456   -79.3754  
6                                  Central Bay Street   43.6564   -79.3860  
7                                            Christie   43.6683   -79.4205  
8                            Richmond, Adelaide, King   43.6496   -79.3833  
9                        Dufferin, Dovercourt Village   43.6655   -79.4378  
10  Harbourfront East, Union Station, Toronto Islands   43.6230   -79.3936  
11                           Little Portugal, Trinity   43.6480   -79.4177  
12                       The Danforth West, Riverdale   43.6803   -79.3538  
13           Toronto Dominion Centre, Design Exchange   43.6469   -79.3823  
14       Brockton, Parkdale Village, Exhibition Place   43.6383   -79.4301  
15                     India Bazaar, The Beaches West   43.6693   -79.3155  
16                     Commerce Court, Victoria Hotel   43.6492   -79.3823  
17                                    Studio District   43.6561   -79.3406  
18                                      Lawrence Park   43.7301   -79.3935  
19                                           Roselawn   43.7113   -79.4195  
20                                   Davisville North   43.7135   -79.3887  
21    Forest Hill North & West, Forest Hill Road Park   43.6966   -79.4120  
22                      High Park, The Junction South   43.6605   -79.4633  
23                  North Toronto West, Lawrence Park   43.7143   -79.4065  
24                The Annex, North Midtown, Yorkville   43.6736   -79.4035  
25                             Parkdale, Roncesvalles   43.6469   -79.4521  
26                                         Davisville   43.7020   -79.3853  
27                     University of Toronto, Harbord   43.6629   -79.3987  
28                                 Runnymede, Swansea   43.6512   -79.4828  
29                        Moore Park, Summerhill East   43.6899   -79.3853  
30          Kensington Market, Chinatown, Grange Park   43.6541   -79.3978  
31  Summerhill West, Rathnelly, South Hill, Forest...   43.6861   -79.4025  
32  CN Tower, King an

### We search for the geographical coordinate of Downtown Toronto.

In [195]:
address = 'Downtown Toronto, TO'

geolocator = Nominatim(user_agent = "DoDo")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Downtown Toronto are 43.6563221, -79.3809161.


In [196]:
print(*zip(boro_toronto['Latitude'], boro_toronto['Longitude'], boro_toronto['Neighbourhood']))

(43.6555, -79.3626, 'Regent Park, Harbourfront') (43.6641, -79.3889, "Queen's Park, Ontario Provincial Government") (43.6572, -79.3783, 'Garden District, Ryerson') (43.6513, -79.3756, 'St. James Town') (43.6784, -79.2941, 'The Beaches') (43.6456, -79.3754, 'Berczy Park') (43.6564, -79.38600000000002, 'Central Bay Street') (43.6683, -79.4205, 'Christie') (43.6496, -79.3833, 'Richmond, Adelaide, King') (43.6655, -79.4378, 'Dufferin, Dovercourt Village') (43.62300000000001, -79.3936, 'Harbourfront East, Union Station, Toronto Islands') (43.648, -79.4177, 'Little Portugal, Trinity') (43.6803, -79.3538, 'The Danforth West, Riverdale') (43.6469, -79.3823, 'Toronto Dominion Centre, Design Exchange') (43.6383, -79.4301, 'Brockton, Parkdale Village, Exhibition Place') (43.6693, -79.3155, 'India Bazaar, The Beaches West') (43.6492, -79.3823, 'Commerce Court, Victoria Hotel') (43.6561, -79.3406, 'Studio District') (43.7301, -79.3935, 'Lawrence Park') (43.7113, -79.4195, 'Roselawn') (43.7135, -79.

In [198]:
# Lets create a map of Downtown Toronto and its surrounding Neighbourhoods
map_dt_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# Showing Downtown Toronto as Red on the map ....
folium.CircleMarker([latitude, longitude], radius=10, popup=label, color='Red', fill=True, fill_color='Red', fill_opacity=0.9,).add_to(map_dt_toronto)

# ... and the Neighbourhoods as Blue.
for lat, lng, label in zip(boro_toronto['Latitude'], boro_toronto['Longitude'], boro_toronto['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        ).add_to(map_dt_toronto)  
    
map_dt_toronto

## Below we Define Foursquare approach by showing what are required before sending out a request or query. One needs the Foursquare Credentials and current Version.

In [230]:
## @ hid
CLIENT_ID = 'xxxxxxxxxxxx' # your Foursquare ID
CLIENT_SECRET = 'xxxxxxxxxxxx' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: xxxxxxxxxxxx
CLIENT_SECRET:xxxxxxxxxxxx


### Foursquare will facilitate getting venues of important places of interest near Downtown Toronto. It will be also possible to obtain relevant information for each nearby venue after sending out request through  Foursquare API

In [200]:
def NearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [201]:
radius = 500
LIMIT = 100 # this will retreive data for only 100 venues per neighbourhood
Boro_Toronto_venues = NearbyVenues(names = boro_toronto['Neighbourhood'], latitudes = boro_toronto['Latitude'], longitudes = boro_toronto['Longitude'])

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

## We can take a look at the dimension of the above venue dataset.

In [203]:
print(Boro_Toronto_venues.shape)
Boro_Toronto_venues.head(15)

(1547, 7)


Neighbourhood  Neighbourhood Latitude  \
0   Regent Park, Harbourfront                 43.6555   
1   Regent Park, Harbourfront                 43.6555   
2   Regent Park, Harbourfront                 43.6555   
3   Regent Park, Harbourfront                 43.6555   
4   Regent Park, Harbourfront                 43.6555   
5   Regent Park, Harbourfront                 43.6555   
6   Regent Park, Harbourfront                 43.6555   
7   Regent Park, Harbourfront                 43.6555   
8   Regent Park, Harbourfront                 43.6555   
9   Regent Park, Harbourfront                 43.6555   
10  Regent Park, Harbourfront                 43.6555   
11  Regent Park, Harbourfront                 43.6555   
12  Regent Park, Harbourfront                 43.6555   
13  Regent Park, Harbourfront                 43.6555   
14  Regent Park, Harbourfront                 43.6555   

    Neighbourhood Longitude                     Venue  Venue Latitude  \
0                  -79.3626             Tandem Coffee       43.653559   
1                  -79.3626          Roselle Desserts       43.653447   
2                  -79.3626    Figs Breakfast & Lunch       43.655675   
3                  -79.3626           The Yoga Lounge       43.655515   
4                  -79.3626       Body Blitz Spa East       43.654735   
5                  -79.3626           Sumach Espresso       43.658135   
6                  -79.3626  Dominion Pub and Kitchen       43.656919   
7                  -79.3626        Morning Glory Cafe       43.653947   
8                  -79.3626                 Sukhothai       43.658444   
9                  -79.3626           Berkeley Church       43.655123   
10                 -79.3626            Rooster Coffee       43.651900   
11                 -79.3626            Impact Kitchen       43.656369   
12                 -79.3626           Alumnae Theatre       43.652756   
13                 -79.3626                 Starbucks       43.651613   
14                 -79.3626    Cooper Koo Family YMCA       43.653249   

    Venue Longitude       Venue Category  
0        -79.361809          Coffee Shop  
1        -79.362017               Bakery  
2        -79.364503       Breakfast Spot  
3        -79.364955          Yoga Studio  
4        -79.359874                  Spa  
5        -79.359515          Coffee Shop  
6        -79.358967                  Pub  
7        -79.361149       Breakfast Spot  
8        -79.365681      Thai Restaurant  
9        -79.365873          Event Space  
10       -79.365609          Coffee Shop  
11       -79.356980           Restaurant  
12       -79.364753              Theater  
13       -79.364917          Coffee Shop  
14       -79.358008  Distribution Center

### From the above code ouput, we can see that there are 1547 venues in the neighbourhoods in this dataset. We can see interesting information about Venue names like Alumnae Theatre, Dominion Pub and Kitchen, Starbucks or Sumach Espresso and their corresponding venue categories like Coffee Shop or Theater.

### Basical we utilize this interesting information to categorize these neigborhoods using k_Means clustering to get insights about business or real estate investments.

### We can examine the number of venues returned per neighbourhood

In [204]:
#pd.set_option('display.max_colwidth', 1)
Boro_Toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Berczy Park                                                             90   
Brockton, Parkdale Village, Exhibition Place                            37   
Business reply mail Processing Centre, South Ce...                      15   
CN Tower, King and Spadina, Railway Lands, Harb...                      57   
Central Bay Street                                                      74   
Christie                                                                12   
Church and Wellesley                                                    74   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              22   
Davisville North                                                         7   
Dufferin, Dovercourt Village                                            18   
First Canadian Place, Underground city                                 100   
Forest Hill North & West, Forest Hill Road Park                          3   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                        4   
High Park, The Junction South                                            3   
India Bazaar, The Beaches West                                          19   
Kensington Market, Chinatown, Grange Park                               64   
Lawrence Park                                                            2   
Little Portugal, Trinity                                                47   
Moore Park, Summerhill East                                              6   
North Toronto West, Lawrence Park                                        4   
Parkdale, Roncesvalles                                                  45   
Queen's Park, Ontario Provincial Government                             28   
Regent Park, Harbourfront                                               22   
Richmond, Adelaide, King                                               100   
Rosedale                                                                 4   
Roselawn                                                                 1   
Runnymede, Swansea                                                      48   
St. James Town                                                          94   
St. James Town, Cabbagetown                                             44   
Stn A PO Boxes                                                          96   
Studio District                                                          8   
Summerhill West, Rathnelly, South Hill, Forest ...                       6   
The Annex, North Midtown, Yorkville                                     20   
The Beaches                                                              8   
The Danforth West, Riverdale                                            36   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          29   

                                                    Neighbourhood Longitude  \
Neighbourhood                                                                 
Berczy Park                                                              90   
Brockton, Parkdale Village, Exhibition Place                             37   
Business reply mail Processing Centre, South Ce...                       15   
CN Tower, King and Spadina, Railway Lands, Harb...                       57   
Central Bay Street                                                       74   
Christie                                                                 12   
Church and Wellesley                                                     74   
Commerce Court, Victoria Hotel                                          100   
Davisville                                                     

## Next we can check how often these unique categories occur.

In [205]:
Boro_Toronto_venues.groupby('Venue Category').count()

Neighbourhood  \
Venue Category                                            
Accessories Store                                     1   
Afghan Restaurant                                     1   
American Restaurant                                  25   
Art Gallery                                          13   
Art Museum                                            1   
Arts & Crafts Store                                   2   
Asian Restaurant                                     15   
Athletics & Sports                                    2   
BBQ Joint                                             3   
Baby Store                                            1   
Bagel Shop                                            1   
Bakery                                               34   
Bank                                                 12   
Bar                                                  23   
Baseball Field                                        1   
Basketball Stadium                                    3   
Beach Bar                                             1   
Beer Bar                                             19   
Beer Store                                            4   
Belgian Restaurant                                    2   
Bistro                                                2   
Bookstore                                            19   
Boutique                                              1   
Bowling Alley                                         1   
Brazilian Restaurant                                  5   
Breakfast Spot                                       25   
Brewery                                               7   
Bubble Tea Shop                                       9   
Building                                              3   
Burger Joint                                          9   
Burrito Place                                         8   
Butcher                                               2   
Café                                                 83   
Camera Store                                          1   
Candy Store                                           2   
Caribbean Restaurant                                  3   
Cheese Shop                                           5   
Chinese Restaurant                                    6   
Chiropractor                                          1   
Clothing Store                                       13   
Cocktail Bar                                         12   
Coffee Shop                                         140   
College Arts Building                                 1   
College Cafeteria                                     1   
College Gym                                           2   
College Rec Center                                    1   
College Theater                                       1   
Colombian Restaurant                                  3   
Comfort Food Restaurant                               4   
Comic Shop                                            2   
Concert Hall                                         10   
Cosmetics Shop                                        7   
Coworking Space                                       1   
Creperie                                              4   
Cuban Restaurant                                      2   
Cupcake Shop                                          1   
Curling Ice                                           1   
Dance Studio                                          5   
Deli / Bodega                                        15   
Department Store                                     10   
Dessert Shop                                         10   
Dim Sum Restaurant                                    1   
Diner                                                13   
Discount Store                                        1   
Distribution Center                                   4   
Dog Run                                               4   
Doner Restaurant              

### From the above we can recognize venue categories that are very outstanding in terms of frequencies or presence. Such venu categories like  American Restaurant(25), Asian Restaurant(15), Bakery(34), Bar(23), Breakfast spot(25), Caffe(83), Coffe Shop(140), Pubs , Hotels, Italian, Japanese and American Restaurant are the most frequently observed venues. The category of that neighbourhood is defined by the occurence of these venues, and we shall make use of these to investigate the cluster of these neighbourhoods.

## Lets check out how many unique categories that exist.

In [207]:
print('There are {} uniques categories.'.format(len(Boro_Toronto_venues['Venue Category'].unique())))

There are 219 uniques categories.


### We can determine the frequency of occurence of each category in a neighborhood
### This requires that, we first convert the venue categories in one-hot vector form such that if a venue category occurs in a particular neighborhood, we assign 1 to that cell(intersection of neighborhood row and venue category column) and 0 for the cells non-occurence.

In [208]:
# one hot encoding
one_hot = pd.get_dummies(Boro_Toronto_venues[['Venue Category']], prefix="", prefix_sep="")
#print(one_hot.head(10))

In [210]:
# add neighborhood column back to dataframe
one_hot['Neighbourhood'] = Boro_Toronto_venues['Neighbourhood'] ## I labeled the new columns as Neighbourhood with a
# 'u' bcs there is already a column named 'Neighborhood' in the venue categories.

# move neighborhood column to the first column
fixed_columns = [one_hot.columns[-1]] + list(one_hot.columns[:-1])
one_hot = one_hot[fixed_columns] 

one_hot.head(20)

Yoga Studio              Neighbourhood  Accessories Store  \
0             0  Regent Park, Harbourfront                  0   
1             0  Regent Park, Harbourfront                  0   
2             0  Regent Park, Harbourfront                  0   
3             1  Regent Park, Harbourfront                  0   
4             0  Regent Park, Harbourfront                  0   
5             0  Regent Park, Harbourfront                  0   
6             0  Regent Park, Harbourfront                  0   
7             0  Regent Park, Harbourfront                  0   
8             0  Regent Park, Harbourfront                  0   
9             0  Regent Park, Harbourfront                  0   
10            0  Regent Park, Harbourfront                  0   
11            0  Regent Park, Harbourfront                  0   
12            0  Regent Park, Harbourfront                  0   
13            0  Regent Park, Harbourfront                  0   
14            0  Regent Park, Harbourfront                  0   
15            0  Regent Park, Harbourfront                  0   
16            0  Regent Park, Harbourfront                  0   
17            0  Regent Park, Harbourfront                  0   
18            0  Regent Park, Harbourfront                  0   
19            0  Regent Park, Harbourfront                  0   

    Afghan Restaurant  American Restaurant  Art Gallery  Art Museum  \
0                   0                    0            0           0   
1                   0                    0            0           0   
2                   0                    0            0           0   
3                   0                    0            0           0   
4                   0                    0            0           0   
5                   0                    0            0           0   
6                   0                    0            0           0   
7                   0                    0            0           0   
8                   0                    0            0           0   
9                   0                    0            0           0   
10                  0                    0            0           0   
11                  0                    0            0           0   
12                  0                    0            0           0   
13                  0                    0            0           0   
14                  0                    0            0           0   
15                  0                    0            0           0   
16                  0                    0            0           0   
17                  0                    0            0           0   
18                  0                    0            0           0   
19                  0                    0            0           0   

    Arts & Crafts Store  Asian Restaurant  Athletics & Sports  BBQ Joint  \
0                     0                 0                   0          0   
1                     0                 0                   0          0   
2                     0                 0                   0          0   
3                     0                 0                   0          0   
4                     0                 0                   0          0   
5                     0                 0                   0          0   
6                     0                 0                   0          0   
7                     0                 0                   0          0   
8                     0                 0                   0          0   
9                     0                 0                   0          0   
10                    0                 0                   0          0   
11                    0                 0                   0          0   
12                    0                 0                   0          0   
13                    0                 0                   0          0   
14    

In [211]:
one_hot.shape

(1547, 220)

### Next we can determine the mean frequency of occurence of each category in a neighborhood

In [213]:
boro_toronto_grouped = one_hot.groupby('Neighbourhood').mean().reset_index()
boro_toronto_grouped.head(10)

Neighbourhood  Yoga Studio  \
0                                        Berczy Park     0.011111   
1       Brockton, Parkdale Village, Exhibition Place     0.000000   
2  Business reply mail Processing Centre, South C...     0.066667   
3  CN Tower, King and Spadina, Railway Lands, Har...     0.017544   
4                                 Central Bay Street     0.000000   
5                                           Christie     0.000000   
6                               Church and Wellesley     0.027027   
7                     Commerce Court, Victoria Hotel     0.000000   
8                                         Davisville     0.000000   
9                                   Davisville North     0.000000   

   Accessories Store  Afghan Restaurant  American Restaurant  Art Gallery  \
0           0.000000           0.000000             0.011111     0.022222   
1           0.027027           0.000000             0.000000     0.027027   
2           0.000000           0.000000             0.000000     0.000000   
3           0.000000           0.000000             0.000000     0.000000   
4           0.000000           0.000000             0.000000     0.000000   
5           0.000000           0.000000             0.000000     0.000000   
6           0.000000           0.013514             0.013514     0.000000   
7           0.000000           0.000000             0.040000     0.010000   
8           0.000000           0.000000             0.000000     0.000000   
9           0.000000           0.000000             0.000000     0.000000   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0    0.000000             0.000000          0.000000            0.000000   
1    0.000000             0.027027          0.000000            0.000000   
2    0.000000             0.000000          0.000000            0.000000   
3    0.000000             0.000000          0.017544            0.000000   
4    0.013514             0.000000          0.000000            0.000000   
5    0.000000             0.000000          0.000000            0.083333   
6    0.000000             0.000000          0.000000            0.000000   
7    0.000000             0.000000          0.030000            0.000000   
8    0.000000             0.000000          0.000000            0.000000   
9    0.000000             0.000000          0.000000            0.000000   

   BBQ Joint  Baby Store  Bagel Shop    Bakery      Bank       Bar  \
0        0.0    0.000000    0.011111  0.044444  0.000000  0.000000   
1        0.0    0.000000    0.000000  0.027027  0.000000  0.027027   
2        0.0    0.000000    0.000000  0.000000  0.066667  0.000000   
3        0.0    0.000000    0.000000  0.035088  0.035088  0.052632   
4        0.0    0.000000    0.000000  0.013514  0.000000  0.000000   
5        0.0    0.083333    0.000000  0.000000  0.000000  0.000000   
6        0.0    0.000000    0.000000  0.000000  0.000000  0.000000   
7        0.0    0.000000    0.000000  0.010000  0.000000  0.010000   
8        0.0    0.000000    0.000000  0.000000  0.000000  0.000000   
9        0.0    0.000000    0.000000  0.000000  0.000000  0.000000   

   Baseball Field  Basketball Stadium  Beach Bar  Beer Bar  Beer Store  \
0             0.0            0.011111   0.000000  0.033333         0.0   
1             0.0            0.000000   0.027027  0.000000         0.0   
2             0.0            0.000000   0.000000  0.000000         0.0   
3             0.0            0.000000   0.000000  0.017544         0.0   
4             0.0            0.000000   0.000000  0.000000         0.0   
5             0.0            0.000000   0.000000  0.000000         0.0   
6             0.0            0.000000   0.000000  0.013514         0.0   
7             0.0            0.000000   0.000000  0.020000         0.0   
8             0.0            0.000000   0.000000  0.000000         0.0   
9             0.0            0.000000   0.000000  0.000000         0.0   

   Belgi

In [214]:
boro_toronto_grouped.shape ### We have 38 unique neighbourhoods

(39, 220)

### Here we print each neighbourhood with top 5 most frequent of its venues.

In [216]:
num_top_venues = 5

for hood in boro_toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = boro_toronto_grouped[boro_toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
         venue  freq
0  Coffee Shop  0.10
1        Hotel  0.06
2         Café  0.04
3       Bakery  0.04
4   Restaurant  0.03


----Brockton, Parkdale Village, Exhibition Place----
                    venue  freq
0                    Café  0.08
1             Coffee Shop  0.08
2          Breakfast Spot  0.05
3  Thrift / Vintage Store  0.05
4               Gift Shop  0.05


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
              venue  freq
0        Restaurant  0.13
1       Coffee Shop  0.13
2       Yoga Studio  0.07
3              Bank  0.07
4  Department Store  0.07


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                venue  freq
0  Italian Restaurant  0.07
1         Coffee Shop  0.07
2                Café  0.05
3                 Bar  0.05
4           Speakeasy  0.04


----Central Bay Street----
              venue  freq
0       Coffee S

### Lets pick top 10 most common venues per neighborhood and use them to create a dataframe

In [217]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [220]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# Labelling columns as 1st, 2nd and so on
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind])) # for 1st, 2nd, 3rd
    except:
        columns.append('{}th Most Common Venue'.format(ind+1)) ### for 4th, 5th,...

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns) ## assign column names we just created to a new dataframe
neighborhoods_venues_sorted['Neighbourhood'] = boro_toronto_grouped['Neighbourhood']## add neighborhoods column

for ind in np.arange(boro_toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(boro_toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head(20)

Neighbourhood 1st Most Common Venue  \
0                                         Berczy Park           Coffee Shop   
1        Brockton, Parkdale Village, Exhibition Place           Coffee Shop   
2   Business reply mail Processing Centre, South C...            Restaurant   
3   CN Tower, King and Spadina, Railway Lands, Har...    Italian Restaurant   
4                                  Central Bay Street           Coffee Shop   
5                                            Christie         Grocery Store   
6                                Church and Wellesley           Coffee Shop   
7                      Commerce Court, Victoria Hotel           Coffee Shop   
8                                          Davisville           Coffee Shop   
9                                    Davisville North               Dog Run   
10                       Dufferin, Dovercourt Village         Grocery Store   
11             First Canadian Place, Underground city           Coffee Shop   
12    Forest Hill North & West, Forest Hill Road Park                  Park   
13                           Garden District, Ryerson           Coffee Shop   
14  Harbourfront East, Union Station, Toronto Islands                  Park   
15                      High Park, The Junction South                  Park   
16                     India Bazaar, The Beaches West                  Park   
17          Kensington Market, Chinatown, Grange Park                  Café   
18                                      Lawrence Park                  Park   
19                           Little Portugal, Trinity           Coffee Shop   

            2nd Most Common Venue                     3rd Most Common Venue  \
0                           Hotel                                      Café   
1                            Café                    Thrift / Vintage Store   
2                     Coffee Shop                               Yoga Studio   
3                     Coffee Shop                                      Café   
4                            Café                        Italian Restaurant   
5                            Café                                      Park   
6                         Gay Bar                       Japanese Restaurant   
7                            Café                                     Hotel   
8                            Café                        Italian Restaurant   
9                  Breakfast Spot                      Gym / Fitness Center   
10                           Park                                    Bakery   
11                           Café                                     Hotel   
12                         Lawyer                                     Trail   
13                 Clothing Store                                      Café   
14                Harbor / Marina                                      Café   
15                  Bowling Alley  Residential Building (Apartment / Condo)   
16           Fast Food Restaurant                                       Pub   
17  Vegetarian / Vegan Restaurant                                       Bar   
18             Photography Studio                                   Dog Run   
19                            Bar                                Restaurant   

    4th Most Common Venue          5th Most Common Venue  \
0                  Bakery             Seafood Restaurant   
1               Gift Shop                 Breakfast Spot   
2                    Bank            Martial Arts School   
3                     Bar                           Park   
4              Restaurant                Bubble Tea Shop   
5              Baby Store                     Playground   
6        Sushi Restaurant                     Restaurant   
7              Restaurant            Japanese Restaurant   
8          Sandwich Place                   Dessert Shop   
9        Department Store              Food & Drink Shop   
10                   Café         Furniture / Home Store   


# Lets see whether we can perform Clustering of the Neighborhoods¶

In [221]:
# set number of clusters
k = 5

boro_toronto_grouped_clustering = boro_toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k, random_state=0).fit(boro_toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:38]

array([3, 3, 3, 3, 3, 0, 3, 3, 3, 0, 0, 3, 4, 3, 0, 0, 3, 3, 1, 3, 0, 0,
       3, 3, 3, 3, 0, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3])

#### We combine all data together, including the cluster labels, neighborhoods and top 10 venues in each neighborhood and venue name.

In [222]:
boro_toronto_merged = boro_toronto

# add clustering labels
boro_toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
boro_toronto_merged = boro_toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

boro_toronto_merged # check the last columns!

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M4E      East Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M6H      West Toronto   
10         M5J  Downtown Toronto   
11         M6J      West Toronto   
12         M4K      East Toronto   
13         M5K  Downtown Toronto   
14         M6K      West Toronto   
15         M4L      East Toronto   
16         M5L  Downtown Toronto   
17         M4M      East Toronto   
18         M4N   Central Toronto   
19         M5N   Central Toronto   
20         M4P   Central Toronto   
21         M5P   Central Toronto   
22         M6P      West Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
25         M6R      West Toronto   
26         M4S   Central Toronto   
27         M5S  Downtown Toronto   
28         M6S      West Toronto   
29         M4T   Central Toronto   
30         M5T  Downtown Toronto   
31         M4V   Central Toronto   
32         M5V  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   

                                        Neighbourhood  Latitude  Longitude  \
0                           Regent Park, Harbourfront   43.6555   -79.3626   
1         Queen's Park, Ontario Provincial Government   43.6641   -79.3889   
2                            Garden District, Ryerson   43.6572   -79.3783   
3                                      St. James Town   43.6513   -79.3756   
4                                         The Beaches   43.6784   -79.2941   
5                                         Berczy Park   43.6456   -79.3754   
6                                  Central Bay Street   43.6564   -79.3860   
7                                            Christie   43.6683   -79.4205   
8                            Richmond, Adelaide, King   43.6496   -79.3833   
9                        Dufferin, Dovercourt Village   43.6655   -79.4378   
10  Harbourfront East, Union Station, Toronto Islands   43.6230   -79.3936   
11                           Little Portugal, Trinity   43.6480   -79.4177   
12                       The Danforth West, Riverdale   43.6803   -79.3538   
13           Toronto Dominion Centre, Design Exchange   43.6469   -79.3823   
14       Brockton, Parkdale Village, Exhibition Place   43.6383   -79.4301   
15                     India Bazaar, The Beaches West   43.6693   -79.3155   
16                     Commerce Court, Victoria Hotel   43.6492   -79.3823   
17                                    Studio District   43.6561   -79.3406   
18                                      Lawrence Park   43.7301   -79.3935   
19                                           Roselawn   43.7113   -79.4195   
20                                   Davisville North   43.7135   -79.3887   
21    Forest Hill North & West, Forest Hill Road Park   43.6966   -79.4120   
22                      High Park, The Junction South   43.6605   -79.4633   
23                  North Toronto West, Lawrence Park   43.7143   -79.4065   
24                The Annex, North Midtown, Yorkville   43.6736   -79.4035   
25                             Parkdale, Roncesvalles   43.6469   -79.4521   
26                                         Davisville   43.7020   -79.3853   
27                     University of Toronto, Harbord   43.6629   -79.3987   
28                                 Runnymede, Swansea   43.6512   -79.4828   
29                        Moore Park, Summerhill East   43.6899   -79.3853   
30          Kensington Market, Chinatown, Grange Park   43.6541   -79.3978   
31  Summerhill West, Rathnelly, South Hill, Forest...   43.6861   

# Lets Visualize the clusters
## We can visualize the clusters of the Neighbourhood categories using Folium library.

In [223]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k)
ys = [i+x+(i*x)**2 for i in range(k)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(boro_toronto_merged['Latitude'], boro_toronto_merged['Longitude'], boro_toronto_merged['Neighbourhood'], boro_toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Lets take a look at the Clusters

In [224]:
# Cluster 0, Red

boro_toronto_merged.loc[boro_toronto_merged['Cluster Labels'] == 0, boro_toronto_merged.columns[[1] + list(range(5, boro_toronto_merged.shape[1]))]] #red

Borough  Cluster Labels 1st Most Common Venue  \
5   Downtown Toronto               0           Coffee Shop   
9       West Toronto               0         Grocery Store   
10  Downtown Toronto               0                  Park   
14      West Toronto               0           Coffee Shop   
15      East Toronto               0                  Park   
20   Central Toronto               0               Dog Run   
21   Central Toronto               0                  Park   
26   Central Toronto               0           Coffee Shop   

   2nd Most Common Venue   3rd Most Common Venue 4th Most Common Venue  \
5                  Hotel                    Café                Bakery   
9                   Park                  Bakery                  Café   
10       Harbor / Marina                    Café           Music Venue   
14                  Café  Thrift / Vintage Store             Gift Shop   
15  Fast Food Restaurant                     Pub      Sushi Restaurant   
20        Breakfast Spot    Gym / Fitness Center      Department Store   
21                Lawyer                   Trail      Doner Restaurant   
26                  Café      Italian Restaurant        Sandwich Place   

     5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
5       Seafood Restaurant                   Beer Bar   Japanese Restaurant   
9   Furniture / Home Store  Middle Eastern Restaurant  Brazilian Restaurant   
10        Doner Restaurant       Fast Food Restaurant        Farmers Market   
14          Breakfast Spot        Japanese Restaurant               Brewery   
15          Sandwich Place         Italian Restaurant           Coffee Shop   
20       Food & Drink Shop                       Park                   Gym   
21       Fish & Chips Shop       Fast Food Restaurant        Farmers Market   
26            Dessert Shop            Thai Restaurant        Farmers Market   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
5             Restaurant          Cocktail Bar            Art Gallery  
9            Pizza Place                   Bar                   Bank  
10    Falafel Restaurant           Event Space   Ethiopian Restaurant  
14     French Restaurant        Sandwich Place             Restaurant  
15            Restaurant           Pizza Place         Ice Cream Shop  
20     Fish & Chips Shop        Farmers Market     Falafel Restaurant  
21    Falafel Restaurant           Event Space   Ethiopian Restaurant  
26  Fast Food Restaurant            Restaurant       Sushi Restaurant

### Most top Venues for cluster label 0 are either Coffe Shop, Sandwich place, Restaurants. We also saw from the map that most neighbourhoods are colored red. In conclusion, Downtown Toronto seems to have swarms of interesting and exciting venues like caffe, Coffee Shop, Amphitheatre and multicultural restaurants.

# Let's examine other clusters

In [225]:
# Cluster 1
boro_toronto_merged.loc[boro_toronto_merged['Cluster Labels'] == 1, boro_toronto_merged.columns[[1] + list(range(5, boro_toronto_merged.shape[1]))]] # purple

Borough  Cluster Labels 1st Most Common Venue  \
18  Central Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18    Photography Studio               Dog Run  Fast Food Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
18        Farmers Market    Falafel Restaurant           Event Space   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
18  Ethiopian Restaurant           Escape Room      Electronics Store

In [227]:
# Cluster 2
boro_toronto_merged.loc[boro_toronto_merged['Cluster Labels'] == 2, boro_toronto_merged.columns[[1] + list(range(5, boro_toronto_merged.shape[1]))]] #blue

Borough  Cluster Labels 1st Most Common Venue  \
27  Downtown Toronto               2                  Café   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
27             Bookstore   Japanese Restaurant                Bakery   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
27           Yoga Studio                  Bank        Sandwich Place   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
27            Restaurant                   Pub      French Restaurant

In [228]:
# Cluster 3
# Airport services 
boro_toronto_merged.loc[boro_toronto_merged['Cluster Labels'] == 3, boro_toronto_merged.columns[[1] + list(range(5, boro_toronto_merged.shape[1]))]] #green

Borough  Cluster Labels  1st Most Common Venue  \
0   Downtown Toronto               3            Coffee Shop   
1   Downtown Toronto               3                    Gym   
2   Downtown Toronto               3            Coffee Shop   
3   Downtown Toronto               3            Coffee Shop   
4       East Toronto               3                    Pub   
6   Downtown Toronto               3            Coffee Shop   
7   Downtown Toronto               3          Grocery Store   
8   Downtown Toronto               3                   Café   
11      West Toronto               3            Coffee Shop   
13  Downtown Toronto               3            Coffee Shop   
16  Downtown Toronto               3            Coffee Shop   
17      East Toronto               3  Performing Arts Venue   
19   Central Toronto               3                    Spa   
22      West Toronto               3                   Park   
23   Central Toronto               3                   Park   
24   Central Toronto               3         Sandwich Place   
25      West Toronto               3                 Bakery   
28      West Toronto               3                   Café   
29   Central Toronto               3                   Park   
30  Downtown Toronto               3                   Café   
31   Central Toronto               3     Light Rail Station   
32  Downtown Toronto               3     Italian Restaurant   
33  Downtown Toronto               3                   Park   
34  Downtown Toronto               3            Coffee Shop   
35  Downtown Toronto               3            Coffee Shop   
36  Downtown Toronto               3            Coffee Shop   
37  Downtown Toronto               3            Coffee Shop   
38      East Toronto               3             Restaurant   

            2nd Most Common Venue                     3rd Most Common Venue  \
0                  Breakfast Spot                                    Bakery   
1                     Coffee Shop                                      Park   
2                  Clothing Store                                      Café   
3                            Café                        Seafood Restaurant   
4                    Neighborhood                                    Bakery   
6                            Café                        Italian Restaurant   
7                            Café                                      Park   
8                     Coffee Shop                                     Hotel   
11                            Bar                                Restaurant   
13                          Hotel                                      Café   
16                           Café                                     Hotel   
17                          Diner                                      Park   
19                  Women's Store                         Food & Drink Shop   
22                  Bowling Alley  Residential Building (Apartment / Condo)   
23                       Gym Pool                                Playground   
24                           Café                            History Museum   
25               Sushi Restaurant               Eastern European Restaurant   
28                    Coffee Shop                               Pizza Place   
29                            Gym                             Grocery Store   
30  Vegetarian / Vegan Restaurant                                       Bar   
31                    Coffee Shop                              Liquor Store   
32                    Coffee Shop                                      Café   
33                  Grocery Store                                Playground   
34                     Restaurant                                     Hotel   
35                     Restaurant                                      Café   
36                           Café                                     Hotel   
37                        Gay Bar                   

In [229]:
# Cluster 4
boro_toronto_merged.loc[boro_toronto_merged['Cluster Labels'] == 4, boro_toronto_merged.columns[[1] + list(range(5, boro_toronto_merged.shape[1]))]] #green

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
12  East Toronto               4      Greek Restaurant           Coffee Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
12        Ice Cream Shop    Italian Restaurant                  Café   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
12            Restaurant           Yoga Studio               Brewery   

   9th Most Common Venue 10th Most Common Venue  
12       Bubble Tea Shop                    Pub